core/templates/dev/head/components/forms/schema_editors/SchemaBasedHtmlEditorDirective.js

```diff
@@ -28,6 +30,200 @@ oppia.directive('schemaBasedHtmlEditor', [
+          // A condition to detect if mutation involves addition of node by
+          // drag and drop method.
+          if (mutations[mutations.length - 1].addedNodes[0]) {
+            // This condition checks if added node is image type.
+            if (mutations[mutations.length - 1]
+              .addedNodes[0].nodeName === 'IMG'){
+              // Gets the added image node
+              var addedImgNode = mutations[mutations.length - 1].addedNodes[0];
+              addedImgNode.classList.add('oppia-noninteractive-image');
+              addedImgNode.classList.add('block-element');
+
+              var MODE_EMPTY = 1;
+              var MODE_SAVED = 3;
+
+              var SOURCE_SEPARATOR = 21;
+              var FILEPATH_SAPERATOR = 48;
+
+              var OUTPUT_IMAGE_FORMAT = 'png';
+              var OUTPUT_IMAGE_MAX_WIDTH_PX = 490;
+
+              $scope.data = {mode: MODE_EMPTY, metadata: {}};
+
+              var getResampledImageData = function(
+                  imageDataURI, width, height) {
+                // Create an Image object with the original data.
+                var img = new Image();
+                img.src = imageDataURI;
+
+                // Create a Canvas and draw the image on it, resampled.
+                var canvas = document.createElement('canvas');
+                canvas.width = width;
+                canvas.height = height;
+                var ctx = canvas.getContext('2d');
+                ctx.drawImage(img, 0, 0, width, height);
+                return canvas.toDataURL('image/' + OUTPUT_IMAGE_FORMAT, 1);
+              };
+
+              var convertImageDataToImageFile = function(dataURI) {
+                // Convert base64/URLEncoded data component to raw binary data
+                // held in a string.
+                var byteString = atob(dataURI.split(',')[1]);
+
+                // Separate out the mime component.
+                var mime = dataURI.split(',')[0].split(':')[1].split(';')[0];
+
+                // Write the bytes of the string to a typed array.
+                var ia = new Uint8Array(byteString.length);
+                for (var i = 0; i < byteString.length; i++) {
+                  ia[i] = byteString.charCodeAt(i);
+                }
+
+                var blob = new Blob([ia], {type: mime});
+                if (blob.type.match('image') &&
+                    blob.size > 0) {
+                  return blob;
+                } else {
+                  return null;
+                }
+              };
+
+              var getTrustedResourceUrlForImageFileName = function(
+                  imageFileName) {
+                var encodedFilepath = window.encodeURIComponent(imageFileName);
+                return $sce.trustAsResourceUrl(
+                  '/imagehandler/' + $scope.explorationId +
+                  '/' + encodedFilepath);
+              };
+
+              $scope.calculateTargetImageDimensions = function() {
```

Comment:
Does this (and the other functions defined on `$scope`) need to be defined inside the Mutation observer code?